datetime/datetime_test.go

```diff
@@ -0,0 +1,543 @@
+	}
+}
+
+func TestCustomEncodeDecodeTuple3(t *testing.T) {
+	if isDatetimeSupported == false {
+		t.Skip("Skipping test for Tarantool without datetime support in msgpack")
+	}
+
+	conn := connectWithValidation(t)
+	defer conn.Close()
+
+	// Setup: insert a value.
+	tm := time.Unix(0, 0)
+	_, err := conn.Insert(space, []interface{}{tm})
+	if err != nil {
+		t.Fatalf("Datetime insert failed: %s", err.Error())
+	}
+
+	var tuples2 []Tuple2
+	err = conn.SelectTyped("testDatetime_2", index, 0, 1, IterEq, []interface{}{1, tm}, &tuples2)
+	if err != nil {
+		log.Fatalf("Failed to SelectTyped: %s", err.Error())
+		return
+	}
+
+	// Teardown: delete a value.
+	_, err = conn.Delete(space, index, []interface{}{tm})
+	if err != nil {
+		t.Fatalf("Datetime delete failed: %s", err.Error())
+	}
+}
+
+func TestCustomDecodeFunction(t *testing.T) {
+	if isDatetimeSupported == false {
+		t.Skip("Skipping test for Tarantool without datetime support in msgpack")
+	}
+
+	conn := connectWithValidation(t)
+	defer conn.Close()
+
+	// Call function 'call_me_maybe' returning a table of custom tuples.
+	var tuples []Tuple
+	err := conn.CallTyped("call_me_maybe", []interface{}{1}, &tuples)
+	if err != nil {
+		t.Fatalf("Failed to CallTyped: %s", err.Error())
+	}
+
+	if cid := tuples[0].Cid; cid != 5 {
+		t.Fatalf("Wrong Cid (%d), should be 5", cid)
+	}
+	if orig := tuples[0].Orig; orig != "Poyekhali!" {
+		t.Fatalf("Wrong Orig (%s), should be 'Hello, there!'", orig)
+	}
+}
+
+func TestCustomEncodeDecodeTuple5(t *testing.T) {
+	if isDatetimeSupported == false {
+		t.Skip("Skipping test for Tarantool without datetime support in msgpack")
+	}
+
+	conn := connectWithValidation(t)
+	defer conn.Close()
+
+	dt := time.Unix(500, 1000)
+	_, err := conn.Insert(space, []interface{}{dt})
+	if err != nil {
+		t.Fatalf("Datetime insert failed: %s", err.Error())
+	}
+
+	resp, errSel := conn.Select(space, index, 0, 1, IterEq, []interface{}{dt})
+	if errSel != nil {
+		t.Errorf("Failed to Select: %s", errSel.Error())
+	}
+	if tpl, ok := resp.Data[0].([]interface{}); !ok {
+		t.Errorf("Unexpected body of Select")
+	} else {
+		if val, ok := tpl[0].(time.Time); !ok || !val.Equal(dt) {
+			t.Fatalf("Unexpected body of Select")
+		}
+	}
+
+	// Teardown: delete a value.
+	_, err = conn.Delete(space, index, []interface{}{dt})
+	if err != nil {
+		t.Fatalf("Datetime delete failed: %s", err.Error())
+	}
+}
+
+// runTestMain is a body of TestMain function
+// (see https://pkg.go.dev/testing#hdr-Main).
+// Using defer + os.Exit is not works so TestMain body
+// is a separate function, see
+// https://stackoverflow.com/questions/27629380/how-to-exit-a-go-program-honoring-deferred-calls
+func runTestMain(m *testing.M) int {
+	isLess, err := test_helpers.IsTarantoolVersionLess(2, 10, 0)
+	if err != nil {
+		log.Fatalf("Failed to extract Tarantool version: %s", err)
+	}
+
+	if isLess {
+		log.Println("Skipping datetime tests...")
+		isDatetimeSupported = false
+		return m.Run()
```

Comment:
Because it is a pattern used in each test of connector that you maintain. See:

```go
./multi/multi_test.go:  return m.Run()
./uuid/uuid_test.go:            return m.Run()
./uuid/uuid_test.go:    return m.Run()
./queue/queue_test.go:  return m.Run()
./bench/go-tarantool/multi/multi_test.go:       return m.Run()
./bench/go-tarantool/uuid/uuid_test.go:         return m.Run()
./bench/go-tarantool/uuid/uuid_test.go: return m.Run()
./bench/go-tarantool/queue/queue_test.go:       return m.Run()
./bench/go-tarantool/connection_pool/connection_pool_test.go:   return m.Run()
./bench/go-tarantool/tarantool_test.go: return m.Run()
./connection_pool/connection_pool_test.go:      return m.Run()
./tarantool_test.go:    return m.Run()
./datetime/datetime_test.go:            return m.Run()
./datetime/datetime_test.go:    return m.Run()
```

I think any ideas regarding refactoring common test code are out of scope of adding datetime support.

Reply:
In fact the pattern has only been used in `uuid/uuid_test.go`:
```go
	isLess, err := test_helpers.IsTarantoolVersionLess(2, 4, 1)
	if err != nil {
		log.Fatalf("Failed to extract tarantool version: %s", err)
	}

	if isLess {
		log.Println("Skipping UUID tests...")
		isUUIDSupported = false
		return m.Run()
	} else {
		isUUIDSupported = true
	}
```

You can `return 0` there too.

Reply:
I see that you skip tests if `isDatetimeSupported == false` in the test cases code. I don't understand why you just don't call `m.Run()` if `isDatetimeSupported == false`.